src/il_representations/algos/pair_constructors.py

```diff
@@ -167,3 +170,81 @@ def __call__(self, data_iter):
                 timestep += 1
+
+
+class JigsawPairConstructor(TargetPairConstructor):
+    def __init__(self, permutation_path, n_tiles=9):
+        self.permutation_path = permutation_path
+        self.permutation = np.load(self.permutation_path)
+        self.n_tiles = n_tiles
+
+    def __call__(self, data_iter):
+        timestep = 0
+        traj_ind = 0
+
+        # Get tile dim
+        first_dict = next(data_iter)
+        original_h, original_w = first_dict['obs'][0].shape[-2], \
+                                 first_dict['obs'][0].shape[-1]
+
+        # We -2 after division here because, as the original paper indicated,
+        # preserving the whole image tile may let the network exploit edge information
+        # and not really learn useful representations. Therefore, here we are using
+        # the central-cropped version of these image tiles.
+        tile_h, tile_w = int(original_h / math.sqrt(self.n_tiles)) - 2, \
+                         int(original_w / math.sqrt(self.n_tiles)) - 2
+
+        permutation_class = [x for x in range(len(self.permutation))]
+        random.shuffle(permutation_class)
+        permute_idx = 0
+
+        for step_dict in data_iter:
+            # Process images into image tiles
+            obs = step_dict['obs']
+            processed_obs_list = []
+            target_list = []
+            for o in obs:
+                processed_obs = self.make_jigsaw_puzzle(o,
+                                                        permutation_class[permute_idx],
+                                                        tile_h,
+                                                        tile_w)
+                processed_obs_list.append(processed_obs)
+            target_list.append(permutation_class[permute_idx])
+            permute_idx = (permute_idx + 1) % len(self.permutation)
+
+            processed_obs_list = np.array(processed_obs_list)
+            target_list = np.array(target_list)
+            yield {
+                'context': processed_obs_list,
+                'target': target_list,
+                'extra_context': [],
+                'traj_ts_ids': [traj_ind, timestep],
+            }
+            if step_dict['dones']:
+                traj_ind += 1
+                timestep = 0
+            else:
+                timestep += 1
+
+    def make_jigsaw_puzzle(self, image, permutation_type, tile_h, tile_w):
+        permute = self.permutation[permutation_type]  # len(permute) = 9
+        n_tiles_sqrt = math.sqrt(self.n_tiles)  # If n_tiles = 9, n_tiles_sqrt = 3
+
+        img_tiles = []
+        unit_h = int(image.shape[0] / n_tiles_sqrt)
+        unit_w = int(image.shape[1] / n_tiles_sqrt)
+
+        for pos in permute:
+            pos_h = int(pos // n_tiles_sqrt) * unit_h
+            pos_w = int(pos % n_tiles_sqrt) * unit_w
+
+            # We +1 after pos_h and pos_w to center crop
+            processed_image = image[pos_h + 1:pos_h + 1 + tile_h, pos_w + 1:pos_w + 1 + tile_w]
```

Comment:
Is the image NCHW or NHWC? This code will only work in the latter case, so I suggest adding an assert to make sure that is the case. I also suggest adding an assert to make sure that `processed_image` has dimensions `tile_h*tile_w*channels`.